### ISSUE TITLE
Socket Exception caused by having too may connections open to ServiceBus topics from binding

### ISSUE THREAD

**Issue by alitaghavi** [Azure/azure-functions-host#3308]
We developed a Function App (2.0 beta) on a dedicated plan (Not Consumption) with 5 instances of S1 server. This function has **IAsyncCollector<Message>** type ServiceBus binding to a topic **with partitioning enabled**. This function continuously publish some results to the topic that comes from the binding at the end of its processing.

After running this function for number times we started getting Socket Exceptions from the ServiceBus. I performed some investigations and I noticed an anomaly number of alive connections in the function app metrics. In the diagram below, the huge spikes of connections happen when the function starts the processing and all the connections remained opened until I restarted the function app the next day.

![fucntion-metric](https://user-images.githubusercontent.com/9065158/44479083-6337ba80-a5f4-11e8-9faf-610a7e4097b4.png)

#### Investigative information

Please provide the following:

- Timestamp: August 13, 2018 - present
- Function App version 2.0-beta
- Region: East US

        [FunctionName("placefunction")]
        public static async Task Run([ServiceBusTrigger("place-commands-queue", Connection = "dhub_imports_adJob_SERVICEBUS")] string adjobCommandsMessage,
            [ServiceBus("events-topic", Connection = "SERVICEBUS_CONN"), ]IAsyncCollector<Message> outputEvents,
            [Blob("projs", FileAccess.ReadWrite, Connection = "STORAGE_CONN")] CloudBlobContainer projectionsStore,
            [Blob("events", FileAccess.ReadWrite, Connection = "STORAGE_CONN")] CloudBlobDirectory eventStore,
            CancellationToken cancellationToken, ILogger log)
        {
            try
            {
              Message result =  new Message(); //the actual code is eliminated but after processing set the result into message
              
              outputEvents.AddAsync(result);
            }
            catch (Exception ex)
            {
                log.LogCritical(ex, ex.Message);
            }
        }

**Comment by paulbatum** [comment]
Hi there, it looks like there is a problem with your code. You made the function async but you are not awaiting the call to .AddAsync(). This means that functions thinks that your function has finished executing and it can start another, while the Add is still in progress. That might be the cause (or it might be unrelated - either way you should fix).

**Comment by alitaghavi** [comment]
Thanks @paulbatum in the code snippet I missed await. I actually have it on my real code. I did a refactor in my code to get around this connection issue. I created my own publisher to my serviceBus topic which I create an static topicClient and reuse it. This resolved my connection issue completely. However, I was expecting the serviceBus binding create a static connection (topicClient) and reuse it during the lifetime of the app.
